### ISSUE TITLE
es6 module (esm)

### ISSUE THREAD

**Issue by konsumer** [RobLoach/node-raylib#135]
Tree-shaking doesn't really work well with commonjs modules, in general, and es6 users (node natively supports esm, now, with `type: module` in package.json) would be able to use our lib better, if it was written as es6 module. We can use other tools (I like microbundle or esbuild) to target lots of different module systems (including very old node commonjs.)

This can help with user's treeshaking on es6-based bundlers, and make native node imports work nicer (currently you have to import all of raylib, not individual functions.) it would also mean we can have esm examples, in this repo, that will allow import and top-level awaits, which would be useful for showing off other addons, like async (#134.)

Microbundle takes slightly longer to build than esbuild, but it's much simpler: make a few changes to package.json that you need to do anyway, to support a bunch of targets, and it will build them all with `microbundle` command.

Related: #130
Related: #134

**Comment by konsumer** [comment]
As an example, we could have demos that look like this:

```js
import {
  BeginDrawing,
  InitWindow,
  BeginDrawing,
  ClearBackground,
  DrawText,
  EndDrawing,
  RAYWHITE,
  LIGHTGRAY
} from 'raylib'
import { MaintainFPS } from 'raylib/addons/async.js'

const screenWidth = 800
const screenHeight = 450

InitWindow(screenWidth, screenHeight, 'raylib [async] example - async game-loop')

while(await MaintainFPS(60)) {
  BeginDrawing()
  ClearBackground(RAYWHITE)
  DrawText('Congrats! You created your first window!', 190, 200, 20, LIGHTGRAY)
  EndDrawing()
}

r.CloseWindow()
```

currently, we can't do mjs in a demo in this repo, but externally (like in a seperate mjs project) you have to do this:

```js
import r from 'raylib'

const {
  BeginDrawing,
  InitWindow,
  BeginDrawing,
  ClearBackground,
  DrawText,
  EndDrawing,
  RAYWHITE,
  LIGHTGRAY
} = r
```

**Comment by RobLoach** [comment]
ES6 also supports the following too...
```
import * as r from 'raylib'
```

I'd be okay with porting everything to es6 modules, but just want to make sure tests still pass. Adopting the new for the sake of adopting the new doesn't actually accomplish much :wink:

**Comment by konsumer** [comment]
> ES6 also supports the following too...
Yep, and you can also just export a default object to keep the same code-style, and other things won't inline that in user's code, if they don't import default and are using a treeshaking bundler.


> I'd be okay with porting everything to es6 modules, but just want to make sure tests still pass. Adopting the new for the sake of adopting the new doesn't actually accomplish much 😉

It's not just to "adopt  the new" for me. A lot of times moving to esm feels like it's just a minor syntax-change, but in reality a ton of stuff actually works way differently, which allows much greater efficiency. Esm builds a graph up-front before running other code, which makes a huge difference in terms of parse-speed, side-effects, and treeshaking. From our perspective, we can just change the generated code to be a little bit different, and through bundling, we don't really need to worry about losing anything, but we would gain being able to operate in esm-space, which has many real benefits.

I'm fine with putting it off, or even never doing it. This project being commonjs doesn't super-effect me ( I still use it in esm projects fine, just have to import the whole thing) except I have to write things in examples/ in CJS, which means no top-level awaits and other little things, but not the end of the world.

**Comment by KTibow** [comment]
At this point, Node 20+ allows for shipping ESM even if consumers are CJS ([source](https://nodejs.org/api/modules.html#loading-ecmascript-modules-using-require#:~:text=v22.0.0,%20v20.17.0)), so no need to bundle or dual-bundle for CJS. node-raylib could go ESM without impacting existing consumers.
